locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -2862,7 +2863,37 @@ protected String getFirstReachableAddress(NodeMetadata node, ConfigBag setup) {
         String result;
         if (enabled) {
             Duration timeout = "true".equals(pollForFirstReachable) ? Duration.FIVE_MINUTES : Duration.of(pollForFirstReachable);
-            result = JcloudsUtil.getFirstReachableAddress(node, timeout);
+
+            Predicate<HostAndPort> pollForFirstReachableHostAndPortPredicate;
+            if (setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE) != null) {
+                LOG.debug("Using pollForFirstReachableAddress.predicate supplied from config for location " + this);
+                pollForFirstReachableHostAndPortPredicate = setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE);
+            } else if (setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE) != null) {
+                LOG.debug("Using pollForFirstReachableAddress.predicate.type supplied from config for location " + this);
+                ConfigBag args = new ConfigBag();
+                final String reachablePredicatePrefix = POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE.getName() + ".";
+                for (Map.Entry<ConfigKey<?>, ?> entry : setup.getAllConfigAsConfigKeyMap().entrySet()) {
+                    String keyName = entry.getKey().getName();
+                    if (keyName.startsWith(reachablePredicatePrefix)) {
+                        args.putStringKey(Strings.removeFromStart(keyName, reachablePredicatePrefix), entry.getValue());
+                    }
+                }
+                pollForFirstReachableHostAndPortPredicate = Reflections.invokeConstructorFromArgs(setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE), new Object[] {args.getAllConfig()}).get();
+            } else {
+                LOG.debug("Using default predicate for pollForFirstReachableAddress for location " + this);
+                pollForFirstReachableHostAndPortPredicate = new Predicate<HostAndPort>() {
```

Comment:
That's right, but you can never be too careful with persistence. Also the same code is used at two other places so makes sense to reuse it.